drake/core/Gradient.h

```diff
@@ -18,55 +18,153 @@ namespace Drake {
+  AutoDiffMatrixType<Derived, Nq> initializeAutoDiff(const Eigen::MatrixBase<Derived> &mat, Eigen::DenseIndex num_derivatives = -1, Eigen::DenseIndex deriv_num_start = 0) {
+    if (num_derivatives == -1)
+      num_derivatives = mat.size();
+
+    AutoDiffMatrixType<Derived, Nq> ret(mat.rows(), mat.cols());
+    initializeAutoDiff(mat, ret, num_derivatives, deriv_num_start);
+    return ret;
+  }
+
+  // TODO begin: move to different location?
+  namespace internal {
+    /** \brief helper for totalSizeAtCompileTime function (recursive)
+     */
+    template<typename Head, typename ...Tail>
+    struct TotalSizeAtCompileTime {
+      static constexpr int eval() {
+        return Head::SizeAtCompileTime == Eigen::Dynamic || TotalSizeAtCompileTime<Tail...>::eval() == Eigen::Dynamic ?
+               Eigen::Dynamic :
+               Head::SizeAtCompileTime + TotalSizeAtCompileTime<Tail...>::eval();
+      }
+    };
+
+    /** \brief helper for totalSizeAtCompileTime function (base case)
+     */
+    template<typename Head>
+    struct TotalSizeAtCompileTime<Head> {
+      static constexpr int eval() {
+        return Head::SizeAtCompileTime;
+      }
+    };
   }
 
-  template <typename Derived>
-  TaylorVecXd constantTaylorVecXd(const Eigen::MatrixBase<Derived>& val) {
-    TaylorVecXd auto_diff_vec(val.rows());
-    constantTaylorVecXd(val, auto_diff_vec);
-    return auto_diff_vec;
+  /** \brief determine the total size at compile time of a number of arguments based on their SizeAtCompileTime static members
+   */
+  template<typename ...Args>
+  constexpr int totalSizeAtCompileTime() {
+    return internal::TotalSizeAtCompileTime<Args...>::eval();
   }
 
+  /** \brief determine the total size at runtime of a number of arguments using their size() methods (base case).
+   */
+  constexpr Eigen::DenseIndex totalSizeAtRunTime() {
+    return 0;
+  }
 
-  /** initTaylorVecXd
-   * @brief initializes the vector with x=value and dx=gradient
+  /** \brief determine the total size at runtime of a number of arguments using their size() methods (recursive)
    */
-  template<typename Derived, typename DerivedGradient>
-  void initTaylorVecXd(const Eigen::MatrixBase<Derived>& val, const Eigen::MatrixBase<DerivedGradient>& gradient, TaylorVecXd& auto_diff_vector)
-  {
-    typedef typename Eigen::MatrixBase<DerivedGradient>::Index Index;
-    auto_diff_vector.resize(gradient.size(),1);
-    auto nx = gradient.cols();
-    for (Index row = 0; row < auto_diff_vector.size(); row++) {
-      auto_diff_vector(row).value() = val(row);
-      auto_diff_vector(row).derivatives().resize(nx,1);
-      auto_diff_vector(row).derivatives() = gradient.row(row).transpose();
-    }
+  template<typename Head, typename ...Tail>
+  Eigen::DenseIndex totalSizeAtRunTime(const Eigen::MatrixBase<Head> &head, const Tail &... tail) {
+    return head.size() + totalSizeAtRunTime(tail...);
+  }
+  // TODO end: move to different location?
```

Comment:
I do think core/ is the right place for these things.  It seems like there are a lot of small utilities like this that we might write around eigen types.  perhaps we need a header file just for EigenUtils.h or something?